### PR TITLE
Support freeze parameters

### LINKED ISSUE #392: Freeze a block for training
## Question
Is there a way to freeze a block for transfer learning ? something similar to Keras **trainable = False** ?

### LINKED ISSUE #1494: Freezing some parameters during backpropagation
Hi,

I would like to know if it is possible to freeze some parameters whose ndarray indexes I want to provide between epochs.
Is is more for experimentation. Based on that, I might want to approach a branch&bound algorithm to optimize the loss value.

Any ideas welcome!

### PR DESCRIPTION
**Pull request by zachgk**
This adds a method to support freezing and unfreezing a parameter for transfer
learning. There is also a helper to (un)freeze all paremters in a block, but
without filtering. For more advanced use cases of (un)freezing part of the
parameters in a block, it should be implemented using block.getChildren(),
block.getDirectParameters() and/or block.getParameters().

fixes #392
fixes #1494

### CODE REVIEW

File: api/src/main/java/ai/djl/nn/Parameter.java [hunk 1]
```diff
@@ -133,6 +133,22 @@ public boolean requiresGradient() {
...
+     *
+     * <p>Sometimes during training, especially during transfer learning, it is typical to train
+     * only part of the model. For this, the freeze can be used to prevent certain parts from being
+     * trained.
+     *
+     * <p>This modifies the {@link #requiresGradient()} of the parameter.
+     *
+     * @param freeze true if the parameter should be frozen ({@code freeze == !requiresGradient()})
+     */
+    public void freeze(Boolean freeze) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/Parameter.java:147]:
```suggestion
    public void freeze(boolean freeze) {
```

File: api/src/main/java/ai/djl/nn/Block.java [hunk 2]
```diff
@@ -268,6 +268,18 @@ default NDList forward(
     void loadParameters(NDManager manager, DataInputStream is)
             throws IOException, MalformedModelException;
 
+    /**
+     * Freezes or unfreezes all parameters inside the block for training.
+     *
+     * @param freeze true if the parameter should be frozen
+     * @see Parameter#freeze(Boolean)
+     */
+    default void freezeParameters(Boolean freeze) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/Block.java:277]:
```suggestion
    default void freezeParameters(boolean freeze) {
```

**Review by frankfliu — APPROVED** [review]
